### PR TITLE
CASSANDRA-16669: Add a toggle to obfuscate password for DCL statements in audit logging, enabled by default

### PR DESCRIPTION
**Pull request by sumanth-pasupuleti**

### CODE REVIEW

File: src/java/org/apache/cassandra/audit/IObfuscator.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.audit;
+
+public interface IObfuscator
+{
+    // Obfuscates the source string and returns the obfuscated string
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/audit/IObfuscator.java]: @sumanth-pasupuleti Would you mind to comment this with "proper" Java doc, not just with one-liner? Thank you very much in advance.

File: src/java/org/apache/cassandra/audit/AuditLogManager.java [hunk 2]
```diff
@@ -114,6 +113,11 @@ private void log(AuditLogEntry logEntry)
     {
         if (!filter.isFiltered(logEntry))
         {
+            // we may need to obfuscate password (CASSANDRA-16669)
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/audit/AuditLogManager.java]: @sumanth-pasupuleti I believe this is not acually necessary, the code is pretty much self-evident.

File: doc/source/new/auditlogging.rst [hunk 3]
```diff
@@ -89,6 +89,7 @@ Audit logging does not log:
 
 1. Configuration changes made in ``cassandra.yaml``
 2. Nodetool Commands
+3. Passwords mentioned as part of DCL statements. Passwords are instead obfuscated as ******
```
Comment (vinaykumarchella) [inline doc/source/new/auditlogging.rst:93]: @sumanth-pasupuleti Can you also update the documentation at `doc/source/operating/audit_logging.rst`, while you are at it, can you add a few example log statements with DCL and password obfuscation at `Sample output` section.

File: test/unit/org/apache/cassandra/audit/PasswordObfuscatorTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Test
+    public void testCreateUserWithoutPassword()
+    {
+        assertEquals("CREATE USER user1", PasswordObfuscator.instance.obfuscate("CREATE USER user1"));
+    }
+
+    @Test
+    public void testAlterUserWithPassword()
```
Comment (vinaykumarchella) [inline test/unit/org/apache/cassandra/audit/PasswordObfuscatorTest.java:91]: Can you add few more tests with different query formats (spaces, newlines, mixed casing at PASSWORD token etc.,), some of these could fail in query validations themselves, but good to have these test cases to validate the regex?

A few examples that I can think of

- ALTER USER user1 with PASSWORD='123' - No spaces, validation error
- ALTER USER user1 with PASSWORD ='123' - valid
- ALTER USER user1 with PASSWORD= '123' - valid
- ALTER USER user1 with PassWoRD = '123' - valid, mixed case
- ALTER USER user1 with passwprd = '123' - valid, small case
- ALTER USER user1 with PassWoRD = '123 - validation error 
- ALTER USER user1 with passwprd = \n '123' - valid, new line? not sure how the cql parser treats these newlines though.

**Comment by smiklosovic** [comment]
@sumanth-pasupuleti it would be awesome if you apply comments of @vinaykumarchella on top of https://github.com/sumanth-pasupuleti/cassandra/pull/3/files

thank you

File: doc/source/new/auditlogging.rst (hunk 3 above)
Comment (sumanth-pasupuleti) [reply doc/source/new/auditlogging.rst:93]: Couldn't quite find doc/source/operating/audit_logging.rst; must have been removed as part of recent website changes I suppose. 
Added a few example dcl statements indicating obfuscated password in the sample output.

File: test/unit/org/apache/cassandra/audit/PasswordObfuscatorTest.java (hunk 4 above)
Comment (sumanth-pasupuleti) [reply test/unit/org/apache/cassandra/audit/PasswordObfuscatorTest.java:91]: I have added these variants into PasswordObfuscatorTest.java

File: doc/source/new/auditlogging.rst (hunk 3 above)
Comment (vinaykumarchella) [reply doc/source/new/auditlogging.rst:93]: Looks like these 2 were merged as part of CASSANDRA-16682.

**Pull request closed without merging**